File: be/landing/src/main/java/com/example/spring_ecom/controller/api/cart/CartAPI.java
package com.example.spring_ecom.controller.api.cart;

import com.example.spring_ecom.controller.api.cart.model.AddToCartRequest;
import com.example.spring_ecom.controller.api.cart.model.CartItemResponse;
import com.example.spring_ecom.controller.api.cart.model.UpdateCartItemRequest;
import com.example.spring_ecom.core.response.ApiResponse;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Tag(name = "Cart", description = "Shopping cart management APIs")
@RequestMapping("/api/cart")
public interface CartAPI {
    
    @Operation(summary = "Get cart items for current user")
    @GetMapping
    ApiResponse<List<CartItemResponse>> getCartItems();
    
    @Operation(summary = "Add item to cart")
    @PostMapping("/items")
    ApiResponse<CartItemResponse> addItemToCart(@Valid @RequestBody AddToCartRequest request);
    
    @Operation(summary = "Update cart item quantity")
    @PutMapping("/items/{productId}")
    ApiResponse<CartItemResponse> updateCartItemQuantity(
            @PathVariable Long productId, 
            @Valid @RequestBody UpdateCartItemRequest request);
    
    @Operation(summary = "Remove item from cart")
    @DeleteMapping("/items/{productId}")
    ApiResponse<Void> removeItemFromCart(@PathVariable Long productId);
    
    @Operation(summary = "Clear all items from cart")
    @DeleteMapping
    ApiResponse<Void> clearCart();

    @Operation(summary = "Sync local cart to server")
    @PostMapping("/sync")
    ApiResponse<Void> syncCart(@RequestBody List<AddToCartRequest> items);
}

File: fe/src/api/cart.api.ts
import apiClient from '@/lib/api-client'
import type { 
  ApiResponse, 
  CartItemResponse, 
  AddToCartRequest, 
  UpdateCartItemRequest 
} from '@/types/api'

const CART_BASE_URL = '/api/cart'

export const cartApi = {
  // Get all cart items
  getCartItems: async (): Promise<CartItemResponse[]> => {
    const response = await apiClient.get<ApiResponse<CartItemResponse[]>>(CART_BASE_URL)
    return response.data.data
  },

  // Add item to cart
  addToCart: async (request: AddToCartRequest): Promise<CartItemResponse> => {
    const response = await apiClient.post<ApiResponse<CartItemResponse>>(
      `${CART_BASE_URL}/items`,
      request
    )
    return response.data.data
  },

  // Update cart item quantity
  updateCartItem: async (productId: number, request: UpdateCartItemRequest): Promise<CartItemResponse> => {
    const response = await apiClient.put<ApiResponse<CartItemResponse>>(
      `${CART_BASE_URL}/items/${productId}`,
      request
    )
    return response.data.data
  },

  // Remove item from cart
  removeCartItem: async (productId: number): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(`${CART_BASE_URL}/items/${productId}`)
  },

  // Clear cart
  clearCart: async (): Promise<void> => {
    await apiClient.delete<ApiResponse<void>>(CART_BASE_URL)
  },

  // Sync local cart to server
  syncCart: async (items: AddToCartRequest[]): Promise<void> => {
    await apiClient.post<ApiResponse<void>>(`${CART_BASE_URL}/sync`, items)
  },
}
